### ISSUE TITLE
Rolling update fails with "Failed to receive 5 resource signal(s)"

### ISSUE THREAD

**Issue by toolmantim** [buildkite/elastic-ci-stack-for-aws#130]
When performing a stack update with running instances it's very common for it to fail, making for a long 50 minutes. #126 was meant to help this, but it appears not really to help at all, but only to make the wait longer.

The stack event log:

```
21:07:41 UTC+10 UPDATE_ROLLBACK_COMPLETE  AWS::CloudFormation::Stack  elastic-runners 
21:07:41 UTC+10 UPDATE_COMPLETE AWS::CloudFormation::Stack  MetricsStack  
21:07:30 UTC+10 DELETE_COMPLETE AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  
21:07:29 UTC+10 UPDATE_IN_PROGRESS  AWS::CloudFormation::Stack  MetricsStack  
21:07:29 UTC+10 DELETE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  
21:07:26 UTC+10 UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS  AWS::CloudFormation::Stack  elastic-runners 
21:07:25 UTC+10 UPDATE_COMPLETE AWS::CloudFormation::Stack  MetricsStack  
21:07:02 UTC+10 UPDATE_COMPLETE AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup 
21:06:58 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-2e3ce11f
21:06:58 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-293ce118
21:06:58 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-3502df04
21:06:58 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-0b77820a
21:06:58 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-c07580c1
21:04:52 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup New instance(s) added to autoscaling group - Waiting on 5 resource signal(s) with a timeout of PT15M.
21:04:52 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Successfully terminated instance(s) [i-2ae73a1b,i-64d52065,i-278b2026,i-73e43942,i-ddd520dc] (Progress 100%).
21:04:50 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Terminating instance(s) [i-2ae73a1b,i-64d52065,i-278b2026,i-73e43942,i-ddd520dc]; replacing with 5 new instance(s).
21:04:49 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Temporarily setting autoscaling group MinSize and DesiredCapacity to 10.
21:04:49 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Rolling update initiated. Terminating 5 obsolete instance(s) in batches of 5. Waiting on resource signals with a timeout of PT15M when new instances are added to the autoscaling group.
21:04:45 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup 
21:04:43 UTC+10 UPDATE_COMPLETE AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  
21:04:42 UTC+10 UPDATE_IN_PROGRESS  AWS::CloudFormation::Stack  MetricsStack  
21:04:36 UTC+10 UPDATE_ROLLBACK_IN_PROGRESS AWS::CloudFormation::Stack  elastic-runners The following resource(s) failed to update: [AgentAutoScaleGroup].
21:04:34 UTC+10 UPDATE_FAILED AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Failed to receive 5 resource signal(s) within the specified duration
20:24:31 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup New instance(s) added to autoscaling group - Waiting on 5 resource signal(s) with a timeout of PT40M.
20:24:30 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Successfully terminated instance(s) [i-47e93476,i-3fed183e,i-6bb61d6a,i-58ea3769,i-4ffa0f4e] (Progress 50%).
20:24:29 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Terminating instance(s) [i-47e93476,i-3fed183e,i-6bb61d6a,i-58ea3769,i-4ffa0f4e]; replacing with 5 new instance(s).
20:24:28 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-278b2026
20:24:28 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-73e43942
20:24:28 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-64d52065
20:24:28 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-2ae73a1b
20:24:06 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Received SUCCESS signal with UniqueId i-ddd520dc
20:20:14 UTC+10 UPDATE_COMPLETE AWS::CloudFormation::Stack  MetricsStack  
20:18:16 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup New instance(s) added to autoscaling group - Waiting on 5 resource signal(s) with a timeout of PT40M.
20:18:15 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Temporarily setting autoscaling group MinSize and DesiredCapacity to 10.
20:18:15 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup Rolling update initiated. Terminating 10 obsolete instance(s) in batches of 5. Waiting on resource signals with a timeout of PT40M when new instances are added to the autoscaling group.
20:18:11 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  AgentAutoScaleGroup 
20:18:04 UTC+10 UPDATE_COMPLETE AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  
20:18:03 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  Resource creation Initiated
20:18:02 UTC+10 UPDATE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration AgentLaunchConfiguration  Requested update requires the creation of a new physical resource; hence creating one.
20:17:56 UTC+10 UPDATE_IN_PROGRESS  AWS::CloudFormation::Stack  MetricsStack  
20:17:48 UTC+10 UPDATE_IN_PROGRESS  AWS::CloudFormation::Stack  elastic-runners User Initiated
```

The ASG log:

```
21:04:57 UTC+10 Successful Launching a new EC2 instance: i-2e3ce11f
21:04:57 UTC+10 Successful Launching a new EC2 instance: i-293ce118
21:04:57 UTC+10 Successful Launching a new EC2 instance: i-3502df04
21:04:57 UTC+10 Successful Launching a new EC2 instance: i-0b77820a
21:04:57 UTC+10 Successful Launching a new EC2 instance: i-c07580c1
21:04:51 UTC+10 Successful Terminating EC2 instance: i-64d52065
21:04:51 UTC+10 Successful Terminating EC2 instance: i-2ae73a1b
21:04:51 UTC+10 Successful Terminating EC2 instance: i-ddd520dc
21:04:51 UTC+10 Successful Terminating EC2 instance: i-73e43942
21:04:50 UTC+10 Successful Terminating EC2 instance: i-278b2026
20:24:29 UTC+10 Successful Terminating EC2 instance: i-3fed183e
20:24:29 UTC+10 Successful Terminating EC2 instance: i-47e93476
20:24:29 UTC+10 Successful Terminating EC2 instance: i-4ffa0f4e
20:24:29 UTC+10 Successful Terminating EC2 instance: i-58ea3769
20:24:29 UTC+10 Successful Terminating EC2 instance: i-6bb61d6a
20:22:00 UTC+10 Successful Launching a new EC2 instance: i-2ae73a1b
20:22:00 UTC+10 Successful Launching a new EC2 instance: i-278b2026
20:22:00 UTC+10 Successful Launching a new EC2 instance: i-73e43942
```

Our code that reports if the instance came up successfully is:
https://github.com/buildkite/elastic-ci-stack-for-aws/blob/e83c9a914039d312656eda5b204870fa6eaa912f/templates/buildkite-elastic.yml#L317-L319

**Comment by toolmantim** [comment]
So it appears this is happening not because new instances aren't marketing themselves as healthy, but that instances aren't being terminated correctly… and then get stuck on `terminating:wait`

The Cloudformation event log has some instances marked as "successfully terminated ..." but in the ASG they're still marked as `Terminating:Wait`. The Cloudformation log then shows `New instance(s) added to autoscaling group - Waiting on 5 resource signal(s) with a timeout of PT40M.` but the ASG only launched 4 instances, and not 5.

SSH'ing into these instances showing `Terminating:Wait` shows no lifecycled log events, but it does show that they failed to run their `cfn-signal` on launch:

```
==> cfn-init.log <==
2016-08-30 01:55:43,893 [INFO] Command 01-write-buildkite-env succeeded
2016-08-30 01:55:43,905 [INFO] Command 01-write-lifecycled-config succeeded
2016-08-30 01:55:54,952 [INFO] Command 02-restart-docker succeeded
2016-08-30 01:55:55,605 [INFO] Command 03-install-buildkite succeeded
2016-08-30 01:55:55,609 [INFO] Test failed with code 1
2016-08-30 01:55:55,612 [INFO] Test failed with code 1
2016-08-30 01:55:55,616 [INFO] ConfigSets completed
2016-08-30 01:55:55,618 [INFO] -----------------------Build complete-----------------------
2016-08-30 01:55:55,805 [DEBUG] CloudFormation client initialized with endpoint https://cloudformation.us-east-1.amazonaws.com
2016-08-30 01:55:55,806 [DEBUG] Signaling resource AgentAutoScaleGroup in stack elastic-runners with unique ID i-cca251d4 and status SUCCESS

==> cfn-wire.log <==
2016-08-30 01:55:43,783 [DEBUG] Response: 200 https://cloudformation.us-east-1.amazonaws.com/?Action=DescribeStackResource&StackName=arn%3Aaws%3Acloudformation%3Aus-east-1%3A445615400570%3Astack%2Felastic-runners%2Fa6408ad0-4c93-11e6-a98f-50d501eb4c17&Version=2010-05-15&ContentType=JSON&LogicalResourceId=AgentLaunchConfiguration [headers: {'x-amzn-requestid': 'dc69d39c-6e54-11e6-a796-a155735a3ad8', 'date': 'Tue, 30 Aug 2016 01:55:43 GMT', 'content-length': '4305', 'content-type': 'application/json'}]
2016-08-30 01:55:55,881 [DEBUG] Response: 400 https://cloudformation.us-east-1.amazonaws.com/?Status=SUCCESS&ContentType=JSON&StackName=elastic-runners&Version=2010-05-15&UniqueId=i-cca251d4&Action=SignalResource&LogicalResourceId=AgentAutoScaleGroup [headers: {'x-amzn-requestid': 'e3bbe9b0-6e54-11e6-a796-a155735a3ad8', 'date': 'Tue, 30 Aug 2016 01:55:55 GMT', 'content-length': '277', 'content-type': 'application/json', 'connection': 'close'}]
2016-08-30 01:55:55,882 [DEBUG] Response error: {"Error":{"Code":"ValidationError","Message":"Stack arn:aws:cloudformation:us-east-1:445615400570:stack/elastic-runners/a6408ad0-4c93-11e6-a98f-50d501eb4c17 is in UPDATE_COMPLETE state and cannot be signaled","Type":"Sender"},"RequestId":"e3bbe9b0-6e54-11e6-a796-a155735a3ad8"}
```

**Comment by toolmantim** [comment]
I just hit this again, where agents were just sitting on `Terminating:Wait` for ages, but they had no `cfn-signal` problems on init… so I don't think it's the init that's the problem. Something is going funky with termination during a rolling update.

**Comment by toolmantim** [comment]
So it appears that when you update the ASG's desired instances manually, or via the scaling policies, the SNS notifications are all sent AOK, and everything works as expected.

But if you do a CF stack update with running instances, SNS terminate notifications don't always seem to be sent, so instances sit there as terminating:wait. This seems to only happen after the first batch of rolling updates, for instances 6 and up.

And if the number of running instances being replaced is more than 10, you'll get a stack update failure.

In a stack with 10 running instances a stack update will finish as UPDATE_COMPLETE while instances were still sitting in terminating:wait.

In a stack with 20 running instances a stack update will fail, as the terminating:wait instances seem to cause the rolling updates to fail because they prevent new instances from spinning up and sending the 5 AOK signals that CF is waiting for.

Still not sure of the root cause, or the fix.

**Comment by toolmantim** [comment]
To recreate:

* Create a stack without metrics, min 0 max 20
* Update the ASG to desired = 20
* Wait for instances to spin up
* Do a stack update that changes the AMI

**Comment by toolmantim** [comment]
It appears this s due to instances not receiving their terminating messages for a very long time.

At the moment we're using a single SQS queue which all instances poll to figure out whether or not they're terminating. This creates some interesting message visibility problems and delays (related: lox/lifecycled#4), because any instance can tie up an SQS message that's destined for another instance.

If 5 out of 50 are terminating you have to try to get the message to those 5… but the other 45 can keep receiving the message and releasing, and there's no guarantees any of the relevant 5 will receive it in a timely fashion.

When scaling down to 0 this problem normally sorts itself out, because as instances are taken out there's more chance of the remaining instances receiving the messages that are destined for them. But if it's a rolling update, with 5 of 30 instances, there's a high chance the instances just never receive their terminating notification within the `PauseTime`.

**Comment by toolmantim** [comment]
Fixed by #135
